### DISCUSSION TITLE
messages flash size are really big for an embedded

### DISCUSSION THREAD

**Discussion by farbodpm** [tokio-rs/prost#606] [General]
I used protocol buffer in C. These days I am trying to use my codes in rust so I am developing code using ptost. Recently I have find out that prost mesages are really big in flash size. I want to take advice about how to use prost or how to optimize my code.

**Comment by LucioFranco** [comment]
You may want to consider using more embedded focused proocols, I don't work much in the embedded space so don't really know the requirements that well but I am sure prost doesn't do much to help that use case that mucn.

**Reply by farbodpm** [reply]
Dear @LucioFranco  ,
Thanks for your response. I have experience in using protobuf in c and that was okay. Now we are migrating from c to rust so that these problems showed up. I was checking prost library last nigh and I realized that the Error handling are so much and so detailed. That's great but a little annoying for embedded systems. May be by adding some optimization the code get much smaller. 
It would be my pleasure to help.

**Comment by farbodpm** [comment]
I also have q question. In a message like below. Ho does the trait of encoding and merge would be created for "all" message?
message A{
 uint32 a = 1;
}
message B{
 uint32 a = 1;
}
message C{
 uint32 a = 1;
}
message all{
  repeated A a =1;
  repeated B b =2;
  repeated C c =4;
}

**Reply by LucioFranco** [reply]
Im not sure what you mean by merge here? Could you explain more?

**Comment by ncs-pl** [comment]
Your problem actually reflects the lack of support for options in Prost!
I think your problems (the main message as well as your question on duplicate generation) could be solved by supporting the [`optimize_for` file option](https://developers.google.com/protocol-buffers/docs/proto3#options).

The `CODE_SIZE` or `LITE_RUNTIME` optimizations would make the code generation more efficient (with, for example, a Reflection-based generation when similar fields are found in various structs).

There are currently open pull requests trying to implement the support for options/custom extensions (both being extremely linked). As for now, you should consider keeping your communication protocols based on Protocol Buffers on C/C++ codes.
